[12.0] spec_driven_model

mileo opened the pull request:
Depends of https://github.com/OCA/l10n-brazil/pull/782 (done)

EDIT by rvalyi: original PR and explanations here: https://github.com/OCA/l10n-brazil/pull/753

rvalyi committed:
[ADD] spec_driven_model module for genDS mixins; WIP

mileo committed:
[WIP] Export xml from spec models

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[NEW] Export button

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[WIP] Export related fields

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[NEW] Export to xml many2one, monetary, datetime fields

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[NEW] Export XML One2many fields

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[REF] Clean code and improve performance

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[REF] Clean code

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[IMP] Float and monetary export

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[WIP] Stacked model recursion

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

gabrielcardoso21 committed:
[ADD] Print item xml

gabrielcardoso21 committed:
[ADD] Export all xml tags from xsd

mileo committed:
[REF] Clean code

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

mileo committed:
[REF] Remove prints

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

gabrielcardoso21 committed:
[REM] Unwanted tags

gabrielcardoso21 committed:
[REM] Empty tags

gabrielcardoso21 committed:
[ADD] Export only correct ICMS tag

gabrielcardoso21 committed:
[ADD] Export CST tag

gabrielcardoso21 committed:
[ADD] Correct export of tag IPI

gabrielcardoso21 committed:
[ADD] Correct export of tag PIS

gabrielcardoso21 committed:
[ADD] Correct export of tag PIS

gabrielcardoso21 committed:
[REM] Tag ISSQN if not service

gabrielcardoso21 committed:
[REF] Show only the correct tags

gabrielcardoso21 committed:
[ADD] Barcode logic

gabrielcardoso21 committed:
[ADD] modBC field

gabrielcardoso21 committed:
[REM] ISSQNtot when there is not any service

gabrielcardoso21 committed:
[REM] Tags II, PISST and COFINSST

mileo committed:
[NEW] ExportDS and ExportXML

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

gabrielcardoso21 committed:
[FIX] Pep8 spec_driven_model

gabrielcardoso21 committed:
[REF] Use _get_concrete function

mileo committed:
[FIX] Replace print with logs and other lint errors

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

gabrielcardoso21 committed:
[FIX] Mensagens do _logger.info

mileo commented:
@rvalyi @renatonlima Os testes estão passando agora.

gabrielcardoso21 committed:
[REM] Logging

gabrielcardoso21 committed:
[ADD] Mensagem no nome fantasia do destinatário para notas em homologação

gabrielcardoso21 committed:
[FIX] Campo valor do produto sem erros devidos a precisão

gabrielcardoso21 committed:
[ADD] Valor correto do campo

gabrielcardoso21 committed:
[ADD] Campo nItem

gabrielcardoso21 committed:
[FIX] Erro na criação das views

mileo committed:
[FIX] pep8

Signed-off-by: Luis Felipe Mileo <mileo@kmee.com.br>

sadamo committed:
[FIX] v['related'] len

gabrielcardoso21 committed:
[REM] Unnecessary comments

gabrielcardoso21 committed:
[REF] Rename function

gabrielcardoso21 committed:
[REF] Encapsulation

gabrielcardoso21 committed:
[REM] Unnecessary comments

gabrielcardoso21 committed:
[REF] Encapsulation

gabrielcardoso21 committed:
[REM] Stacked from res_partner and res_company

gabrielcardoso21 committed:
[FIX] Pep8

rvalyi commented:
(branch is out of date, explanations about the concept can be found here https://github.com/OCA/l10n-brazil/pull/753 )

gabrielcardoso21 committed:
[FIX] Import

gabrielcardoso21 committed:
[REF] One2many export

mileo commented:
@rvalyi você esta trabalhando neste módulo em outra branch? 

Podemos revisar este aqui e proceder com o merge? Os testes estão passando desde 30 de abril....

Fizemos mais melhorias aqui: https://github.com/kmee/l10n-brazil/tree/12.0.2.0-staging/spec_driven_model

@gabrielcardoso21 pode colocar as melhorias aqui?

rvalyi commented:
@mileo, sim tou trabalhando na branch que extraimos da branch da nfe do odoo-brazil antes de vcs começarem a botar novas funcionalidades. E realmente essa questao de vcs botar funcionalidades nas mesmas branches antes de cuidar de deixar a base limpa para a OCA nao foi um ponto isolado Luis. Alem da DFe, vc tb tava partindo para enviar wizards de importaçao de NFe que ia quebrar o framework de mapping e ate o modelo relacional no fiscal.

para prototipar e para mostrar que funciona fica mais facil ter os modulos na mesma branch e depois extrair o modulo no PR. Ja vou subir o que trabalhei. Mas ontem deixamos os modulos l10n_br_purchase e l10n_br_purchase_stock verde para review (PR de dezembro https://github.com/OCA/l10n-brazil/pull/757 ) entao ja da para vcs revisarem isso enquanto isso. Se subir uns commits de export aqui vou ver para integrar, depois pode ser feito depois.

A ideia e mostrar logo que a limpeza do l10n_br_nfe_spec ta funcionando e melhor, fazer o merge dele e depois extrair o spec_driven_model de novo e cuidar do merge dele.

rvalyi commented:
> Podemos revisar este aqui e proceder com o merge? Os testes estão passando desde 30 de abril....

Nao. Tem uma mudanças importantes para chegar. Sobre os testes passando nesse modulo, logico que passa, quase nada ta testado aqui, assim é mole. Acontece que esse tipo de modulo é bem complexo de testar e a melhor forma de testar que temos é o l10n_br_nfe por enquanto no qual tou testando um pouco tudo junto. E nisso da para entender o procedimento: limpar primeiro a geraçao dos mixin com generateDS, limpar o nfelib para botar pos pedaços juntos. E mole fazer as coisas de qualquer jeito no neflib e generateDS e depois um modulo desse passar o flake8 e falar que ta okay para merge, mas nao é razoavel. Ai enfim como vc tava fazendo pressao para outras coisas do que limpando as bases da NFe tb cuidei dos outros assuntos antes para nao sabotar o que ja temos na OCA.

Mas ainda assim pretendo subir uns testes no spec_model_driven antes de rolar merge.

rvalyi commented:
E novamente Luis, se a KMEE tivesse feito esse tipo de trabalho de limpeza da NFe uns 5 anos atras em vez de partir para folha ou SPED ou PDV que na hora do vamos ver, tem nada que tem qualidade de OCA, certamente tudo isso teria sido feito antes. Procura um responsável? Olha no espelho primeiro.

rvalyi committed:
avoids logger error and logs StackModel init

rvalyi committed:
spec.mixin dynamically injected -> no dep required

rvalyi committed:
removed most of NFe schema specifics

rvalyi committed:
skip int/float/monetary defaults because None->0

rvalyi committed:
assume XSD tags can be skipped from generated models

rvalyi committed:
better dry-run support (no create)

rvalyi committed:
propagate path in _build_many2one for better import control

mileo commented:
@rvalyi pode fazer um rebase a questão do openssl foi corrigida.

rvalyi committed:
normal py3 StringIO import

rvalyi commented:
@mileo. Fiz o rebase.
@mileo @gabrielcardoso21 @renatonlima @marcelsavegnago @clementmbr . Eu estou finalizando os commits nesse PR para ter os testes de NFe bem certinho com modelo de dado limpo. Devo encerrar essa tarde. Tou testando uns detalhes a fundo antes de fazer os commits (eu tinha alguns hacks para passar os testes). Ja vou subir um PR com rebase da Nfe nisso e testes passando, antes das novas funcionalidades, ai depois a gente deveria fazer uns cherry-picks das coisas de NFe que o Gabriel e o Renato fizeram recentemente.

Novamente a questao é que esse modulo é bem complexo de testar. O melhor teste que temos hoje esta no l10n_br_nfe. Porem, foi acumulado funcionalidades no l10n_br_nfe na branch da KMEE antes de pensar em acertar a qualidade minima/modelo de dados e isso atrapalha para acertar essa qualidade, teremos que re-adicionar essas funcionalidades com cherry-pick depois. Pretendo botar testes independentes spec_model_driven, especialmente antes de propor ele num repo como OCA/server-tools na 14. Porem com os testes legais no l10n_br_nfe accredito que poderia rolar merge aqui antes.

Uma consideraçao importante: era ruim de ter os modulos de mixins geridos depender desse modulo spec_model_driven no PR #753 que eu tinha proposto inicialmente, pois esse modulo spec_model_driven tem bastante opinioes e pode nao se concordar com tudo, enquantos mixins Odoo geridos que reflitam fielmente un XSD tem pouco debate.

Entao nesse PR https://github.com/OCA/l10n-brazil/pull/981, o l10n_br_nfe_spec nao depende mais do spec_model_driven e é entao mais simples de avaliar/adotar. 2 commits importantes para conseguir essa independencia sao esses:

- https://github.com/OCA/l10n-brazil/pull/789/commits/0bd90fbe5bb051dc8446dceebe363711f1e8198e
- https://github.com/OCA/l10n-brazil/pull/789/commits/652c9156fe00896a16df28693f53b83dcc368023

rvalyi committed:
important: map_concrete spec classes injected in SpecModel

rvalyi committed:
class _prepare_import_dict hook when m2o are created

rvalyi commented:
uma nota importante, sobre esse commit cef12c0 de hoje (em cima). Isso permite que com o modulo l10n_br_nfe, as tabela geridas de forma concreta para as tags do XSD da nfe 4.0 sejam apenas essas 46:

```
 public | nfe_40_adi                                              
 public | nfe_40_arma                                           
 public | nfe_40_autxml                                        
 public | nfe_40_avulsa                                         
 public | nfe_40_cana                                           
 public | nfe_40_card                 
 public | nfe_40_cide                  
 public | nfe_40_cobr                 
 public | nfe_40_compra            
 public | nfe_40_deduc              
 public | nfe_40_detexport         
 public | nfe_40_detpag             
 public | nfe_40_di                     
 public | nfe_40_dup                  
 public | nfe_40_encerrante       
 public | nfe_40_exporta            
 public | nfe_40_exportind         
 public | nfe_40_fat                    
 public | nfe_40_fordia               
 public | nfe_40_impostodevol    
 public | nfe_40_infadic              
 public | nfe_40_infnfesupl         
 public | nfe_40_infprot              
 public | nfe_40_infrec               
 public | nfe_40_ipi                    
 public | nfe_40_lacres               
 public | nfe_40_nfref                 
 public | nfe_40_obscont            
 public | nfe_40_obsfisco            
 public | nfe_40_procref             
 public | nfe_40_rastro               
 public | nfe_40_refecf               
 public | nfe_40_refnf                 
 public | nfe_40_refnfp               
 public | nfe_40_rettransp          
 public | nfe_40_tconsrecinfe     
 public | nfe_40_tenvinfe            
 public | nfe_40_tinfresptec        
 public | nfe_40_tnfe                  
 public | nfe_40_tnfeproc           
 public | nfe_40_tprotnfe            
 public | nfe_40_transporta        
 public | nfe_40_tretconsrecinfe 
 public | nfe_40_tretenvinfe       
 public | nfe_40_tveiculo            
 public | nfe_40_vol  
```

antes desse commit considerando essa herança no res.partner:

```
class ResPartner(spec_models.SpecModel):
    _name = 'res.partner'
    _inherit = ['res.partner', 'nfe.40.tendereco',
                'nfe.40.tlocal', 'nfe.40.dest']
```

antes do commit, o modulo ia criar essas tabelas de forma inutil (alem de outras pegadinhas):
```
      public | nfe_40_dest          
      public | nfe_40_tendereco
      public | nfe_40_tlocal         
```
Era inutil ja que a tabela res_partner esta usada para essas informações com um mapping apropriado.

cc @renatonlima @clementmbr @marcelsavegnago @gabrielcardoso21 @mileo

rvalyi committed:
_inject_spec_mixin before super -> less loops inside method

rvalyi committed:
document spec model attributes without inconsistently shadowing spec.mixin.<schema_name> attributes

rvalyi committed:
flake8 + comment

mileo commented:
@rvalyi já pode revisar? Vc ainda vai trabalhar nele? Quer alguma ajuda?

rvalyi commented:
> @rvalyi já pode revisar? Vc ainda vai trabalhar nele? Quer alguma ajuda?

Entao, aqui tou finalizando esse PR https://github.com/OCA/l10n-brazil/pull/1043 para passar os testes minimalmente de novo concertando a importaçao sistematica (sobre qual a importaçao com deparo manual por ser feita correctamente) e usando um l10n_br_nfe_spec criado de forma limpa e que nao bagunça o modelo de dados como tava no caso das branches antigas, alem de realmente criar muitas dores de cabeça com as classes e campos de ICMS** e ICMSSN** que tb criava muitos conflitos de campos no Odoo: https://github.com/OCA/l10n-brazil/pull/981

Meu modulo spec_model_driven ta quasse igual a dessa branch aqui localmente, mas me da mais essa noite para checar o spec_model_driven ainda arrumando os testes no l10n_br_nfe (eu tinha eles passando, porem separando os commits me deparei com um hack feio aqui ou la e tou limpando isso subindo os commits no PR #1043 

Como vcs podem ver, esse PR do spec_model_driven ta quase igual ao da branch de nfe que a KMEE usa entao eu accredito que esse merge poderia rolar logo (no fds?).

Ja o PR do l10n_br_nfe_spec https://github.com/OCA/l10n-brazil/pull/981 ta um pouco diferente (na verdade muito mais limpo, menor e com testes razoaveis), mas ele vai depender desse PR https://github.com/erpbrasil/erpbrasil.edoc/pull/15 (que vcs reverteram mas vou refazer) e realmente vai exigir as minha ultimas mudanças no l10n_br_nfe alem desse PR aqui para funcionar igual antes. Nisso eu nao aconselho de fazer o PR do l10n_br_nfe_spec tao logo, digamos antes de ter os testes passando legal no #1043 e vcs da KMEE analisando o diff com a branch de vcs ao fundo. Porque? Porque senao vai criar conflito com o l10n_br_nfe_spec da branch de vcs quando vcs vao tentar dar merge e ai vai dar muita dor de cabeça para vcs (o mesmo acontece com o que a gente ta usando na prod pro enquanto ta).

cc @gabrielcardoso21 @marcelsavegnago @renatonlima

rvalyi commented:
talvez tem mais umas 4-5 linhas que irei precisar alterar, mas realmente acabei de concertar os testes com l10n_br_nfe locamente (vou subir logo) nao deve mudar muita coisa. Nisso ja pode ser revisado sim.

rvalyi committed:
overrideable create_m2o in _prepare_import_dict

useful for importing NFe emit as dest and avoiding to create
a broken related enderEmit for enderDest.

rvalyi committed:
license headers -> LGPL 3

rvalyi commented:
in commit 8c70a07 I put the missing license headers. @gabrielcardoso21 @mileo you contributed the spec_export code under AGPL, Is it ok for you to change that to LGPL before it's released? In the future I would like to put this module under OCA/server-tools and I think it's going to be more popular if it's LGPL as many "framework" tools. As a reminder Akretion was pushing AGPL like mad in the OCA back in 2014 because we believed no word from Fabien that the Odoo license change was for the good of the developpers for them selling their apps on the appstore. We knew since teh beginning that it was a divide and conquer strategy and that the hideen agenda was a transition to an open core business model. We wanted to get the OCA make an AGPL block to make pressure have a chance to avoid this relicensing. But now that it's over on Odoo SA side, despite 90% + of Akretion contribs to the OCA are still AGPL, we better be pragamatic and more fexible with that, specially on the framework side (like this one) or small module that could be used in Odoo EE eventually (se we don't see a competition just because of the license).

cc @mileo @renatonlima

rvalyi commented:
@gabrielcardoso21 @mileo @marcelsavegnago @renatonlima importante: usando o spec_model_driven nessa forma, consigo tirar umas 200 linhas de merda de mapping com IF e isso so no document.py e sem nem considerar as coisas que o Gabriel accrescentou desde entao: https://github.com/akretion/l10n-brazil/commit/942e1d5b4c838f20ccd637b5efaaa0fc96906438
(mesmo que ainda nao fiz commit dos testes, tou testando que importa certinho e sai igual ta).
Agora vou separar os commits de limpeza no document_line.py e botar os testes completos.

mileo commented:
> in commit [8c70a07](https://github.com/OCA/l10n-brazil/commit/8c70a07492a831eeabb9d3ac8ea286e7962d11b1) I put the missing license headers. @gabrielcardoso21 @mileo you contributed the spec_export code under AGPL, Is it ok for you to change that to LGPL before it's released? In the future I would like to put this module under OCA/server-tools and I think it's going to be more popular if it's LGPL as many "framework" tools. As a reminder Akretion was pushing AGPL like mad in the OCA back in 2014 because we believed no word from Fabien that the Odoo license change was for the good of the developpers for them selling their apps on the appstore. We knew since teh beginning that it was a divide and conquer strategy and that the hideen agenda was a transition to an open core business model. We wanted to get the OCA make an AGPL block to make pressure have a chance to avoid this relicensing. But now that it's over on Odoo SA side, despite 90% + of Akretion contribs to the OCA are still AGPL, we better be pragamatic and more fexible with that, specially on the framework side (like this one) or small module that could be used in Odoo EE eventually (se we don't see a competition just because of the license).
> 
> cc @mileo @renatonlima

Por mim sem problemas.

mileo commented on spec_driven_model/hooks.py:
```diff
@@ -0,0 +1,121 @@
+    access_data = []
+    for model in remaining_models:
+        underline_name = model.replace('.', '_')
+        rec_id = "acl_%s_nfe_40_%s" % ('todo'.split('.')[-1],
+                                       underline_name)
+        # TODO no nfe ref
+        model_id = "l10n_br_nfe_spec.model_%s" % (underline_name,)
+        access_data.append([rec_id, underline_name, model_id,
+                            'base.group_user', '1', '1', '1', '1'])
+        # TODO make more secure!
```
Creio que este nfe_40 e l10n_br_nfe_spec não deveriam estar hardcoded

mileo commented on spec_driven_model/hooks.py:
```diff
@@ -0,0 +1,121 @@
+
+    # visit_stack will now need the associated spec classes
+    injected_classes = set()
+    remaining_models = set(['nfe.40.tveiculo'])  # TODO
```
Novamente aqui

mileo commented on spec_driven_model/models/spec_export.py:
```diff
@@ -0,0 +1,179 @@
+
+    def _get_ds_class(self, class_obj):
+        #  FIXME: leiauteNFe hardcoded
+        return getattr(leiauteNFe, class_obj._generateds_type)
```
Aqui tb.

mileo commented on spec_driven_model/models/spec_export.py:
```diff
@@ -0,0 +1,179 @@
+        for c in set(classes):
+            if c is None:
+                continue
+            if 'nfe.' not in c:  # make generic brittle
```
Aqui tb.

mileo commented on spec_driven_model/models/spec_import.py:
```diff
@@ -0,0 +1,270 @@
+        value = getattr(node, attr.get_name())
+        if value is None or value == []:
+            return False
+        key = "nfe40_%s" % (attr.get_name(),)  # TODO schema wise
```
Aqui tb.

mileo commented on spec_driven_model/models/spec_import.py:
```diff
@@ -0,0 +1,270 @@
+            else:
+                clean_type = attr.get_child_attrs()[
+                    'type'].replace('Type', '').lower()
+                comodel_name = "nfe.40.%s" % (clean_type,)  # TODO clean
```
Aqui tb.

mileo commented on spec_driven_model/models/spec_import.py:
```diff
@@ -0,0 +1,270 @@
+        fields = self.fields_get()
+        for k, v in fields.items():
+            # select schema choices for a friendly UI:
+            if k.startswith('nfe40_choice'):  # TODO schema wise
```
Aqui tb.

mileo commented on spec_driven_model/models/spec_view.py:
```diff
@@ -0,0 +1,323 @@
+
+_logger = logging.getLogger(__name__)
+
+TAB_NAME = "NFe"  # TODO dynamic
+FIELD_PREFIX = "nfe40_"
+CHOICE_PREFIX = "nfe40_choice"
+# TODO use schema name as default root key unless a key is provided in the
```
Aqui tb

gabrielcardoso21 reviewed (approved):
Pelo que testei, essas modificações não quebram nada do que já está funcionando na branch que estamos usando.

mileo commented:
@rvalyi fiz alguns comentário, mas isso já deve estar no seu Roadmap.

marcelsavegnago commented:
> Pelo que testei, essas modificações não quebram nada do que já está funcionando na branch que estamos usando.

@gabrielcardoso21 Sábado eu estava testando com as PRs que o Raphael está modificando e tive problema aqui:

/spec_import.py", line 111, in _build_attr
    create_m2o, value, child_path)
TypeError: _build_many2one() takes 6 positional arguments but 8 were given

usei estas PRs 
    - oca refs/pull/789/head
    - oca refs/pull/981/head
    - oca refs/pull/1043/head

Obs. tbm utilizei o respositório nfelib da akretion e com esta PR https://github.com/erpbrasil/erpbrasil.edoc/pull/15

Você conseguiu instalar o módulo l10n_br_nfe?

gabrielcardoso21 commented:
> > Pelo que testei, essas modificações não quebram nada do que já está funcionando na branch que estamos usando.
> 
> @gabrielcardoso21 Sábado eu estava testando com as PRs que o Raphael está modificando e tive problema aqui:
> 
> /spec_import.py", line 111, in _build_attr
> create_m2o, value, child_path)
> TypeError: _build_many2one() takes 6 positional arguments but 8 were given
> 
> usei estas PRs
> - oca refs/pull/789/head
> - oca refs/pull/981/head
> - oca refs/pull/1043/head
> 
> Obs. tbm utilizei o respositório nfelib da akretion e com esta PR [erpbrasil/erpbrasil.edoc#15](https://github.com/erpbrasil/erpbrasil.edoc/pull/15)
> 
> Você conseguiu instalar o módulo l10n_br_nfe?

Eu testei apenas substituindo o spec_driven_model da branch 12.0-l10n_br_nfe pelo módulo desse PR. Foi apenas necessário substituir os campos dos modelos que antes estavam hardcoded. A revisão dos PRs posteriores vai ser mais demorada, pois os testes não podem ser feitos simplesmente plugando o módulo na solução atual. Porém, não vejo motivo para impedir o merge desse PR.

mileo commented:
/ocabot merge nobump

Pessoal para ficar mais fácil de avançar com os outros PRs estou solicitando o merge.

Abrimos um issue com as melhorias a serem feitas nesse módulo e/ou continuamos a discussão no próximo PR.
